### PR TITLE
feat: Func plugin to add/remove functions within function app

### PR DESCRIPTION
**Pull request by tbarlow12**
- Adds CLI commands `sls func add -n {functionName}` and `sls func remove -n {functionName}`
- Adds 1 plugin with 3 hooks: (`func`, `func:add`, `func:remove`)
- Generates code directory and updates `serverless.yml` appropriately
- For now, only supports http functions. Working on supporting additional triggers
- Still working on tests, documentation and edge cases

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -4,16 +4,20 @@ This way only one plugin needs to be added to the service in order to get access
...
-import AzureProvider from './provider/azureProvider';
-import { AzureInvoke } from './plugins/invoke/azureInvoke';
-import { AzureLogs } from './plugins/logs/azureLogs';
-import { AzureRemove } from './plugins/remove/azureRemove';
-import { AzurePackage } from './plugins/package/azurePackage';
-import { AzureDeployPlugin } from './plugins/deploy/azureDeployPlugin';
-import { AzureLoginPlugin } from './plugins/login/loginPlugin';
-import { AzureApimServicePlugin } from './plugins/apim/apimServicePlugin';
-import { AzureApimFunctionPlugin } from './plugins/apim/apimFunctionPlugin';
+import Serverless from "serverless";
```
Comment (tbarlow12) [inline src/index.ts]: Updated to use double quotes

File: src/plugins/func/funcUtils.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+import yaml from "js-yaml";
+import fs from "fs";
+
+const functionsRegex = /functions:([\s\S]*?)\n\n/g
```
[the remark this answers is not in the document]
Comment (wbreza) [reply src/plugins/func/funcUtils.ts]: Does this go back to preserving comments?  From my investigation this library doesn't support comments and don't plan on it.  Do we "really" care about preserving comments?

[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/plugins/func/funcUtils.ts]: The easiest way would be to just parse the yaml, update the functions section and write it back. I'm fine with that if you're ok with it

File: src/plugins/func/azureFunc.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import Serverless from "serverless";
+
+export class AzureFuncPlugin {
```
[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/plugins/func/azureFunc.ts]: This was part of my investigation into sub-commands. But you're right, I think I can unify these at this point

File: src/plugins/func/remove/azureFuncRemove.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import fs from "fs";
+import rimraf from "rimraf";
+import Serverless from "serverless";
+import { FuncPluginUtils } from "../funcUtils";
+
+export class AzureFuncRemovePlugin {
```
[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/plugins/func/remove/azureFuncRemove.ts]: Same as below

File: src/plugins/func/funcUtils.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        };
+    }
+};`;
+  }
+
+  public static getFunctionJson(name: string, options: any) {
+    // TODO: This is where we would just generate function JSON from SLS object
+    // using getFunctionSlsObject(name, options). Currently defaulting to http in and out
+    return JSON.stringify({
+      "bindings": [
```
[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/plugins/func/funcUtils.ts]: Fine with me. Maybe have a folder of templates as we start to support other bindings. I'm envisioning a `func update` command as well

File: src/plugins/func/remove/azureFuncRemove.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/plugins/func/remove/azureFuncRemove.ts]: Combined in update

File: src/plugins/func/azureFunc.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/plugins/func/azureFunc.ts]: Combined in update

File: src/plugins/func/funcUtils.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+  public static updateFunctionsYml(functionYml: any, serverlessYml?: string) {
+    serverlessYml = serverlessYml || FuncPluginUtils.getServerlessYml();
+    const newFunctionsYaml = yaml.dump(functionYml);
+    const newServerlessYaml = serverlessYml.replace(functionsRegex, `${newFunctionsYaml}\n`);
+    fs.writeFileSync("serverless.yml", newServerlessYaml);
+  }
+
+  public static getFunctionHandler(name: string) {
+    return `"use strict";
```
[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/plugins/func/funcUtils.ts]: So I've spent some time trying to make this work. My problem has been what kind of file to put the template in. Here are the options I've tried:

Template file options:
1. TypeScript file
   - Gets transpiled to JavaScript in build - doesn't keep template structure
2. JavaScript file
   - Can't be included in build
   - Tried setting 'allowJs' to be true, but that conflicts with 'declarations' being true
3. Text file
   - Could make it work, but it would mean an extra step post-build of copying the text file over. Not sure that's warranted just to avoid the inline script

Other options?

File: src/plugins/func/azureFunc.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  private async func() {
+    this.serverless.cli.log("Use the func plugin to add or remove functions within Function App");
+  }
+
+  private async add() {
+    if (!("name" in this.options)) {
+      this.serverless.cli.log("Need to provide a name of function to add");
+      return;
```
Comment (mydiemho) [inline src/plugins/func/azureFunc.ts]: if you return instead of throwing error, would this exit other lifecycle?

File: src/plugins/func/azureFunc.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+      this.serverless.cli.log(`Function ${funcToAdd} already exists`);
+      return;
+    }
+    this.createFunctionDir(funcToAdd);
+    this.addToServerlessYml(funcToAdd);
+  }
+
+  private createFunctionDir(name: string) {
+    this.serverless.cli.log("Creating function dir");
+    fs.mkdirSync(name);
```
Comment (mydiemho) [inline src/plugins/func/azureFunc.ts]: should we worry about error here? in cases of permission issues?

File: src/plugins/func/funcUtils.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    return JSON.stringify(httpBinding);
+  }
+
+  public static getFunctionSlsObject(name: string, options: any) {
+    return FuncPluginUtils.defaultFunctionSlsObject(name);
+  }
+
+  private static defaultFunctionSlsObject(name: string) {
+    return {
+      handler: `${name}/index.handler`,
```
Comment (mydiemho) [inline src/plugins/func/funcUtils.ts]: we don't want `${name}/index.handler`, this will generate incorrect binding and function won't work.  

just `index.handler`

File: src/test/mockFactory.ts [hunk 10]
```diff
@@ -37,6 +39,50 @@ export class MockFactory {
...
+    for (let i = 0; i < functionCount; i++) {
+      const functionName = `function${i+1}`;
+      data[functionName] = MockFactory.createTestFunctionMetadata(functionName)
+    }
+    return (wrap) ? {"functions": data } : data;
+  }
+
+  public static createTestFunctionMetadata(name: string) {
+    return {
+      "handler": `${name}/index.handler`,
```
Comment (mydiemho) [inline src/test/mockFactory.ts]: same as above no `${name}`

File: src/plugins/func/funcUtils.ts (hunk 9 above)
Comment (tbarlow12) [reply src/plugins/func/funcUtils.ts]: This is the structure that `serverless.yml` needs to find the handler since it will live inside the directory `{functionName` and inside the file `index.js`, with the function name `handler`

File: src/test/mockFactory.ts (hunk 10 above)
Comment (tbarlow12) [reply src/test/mockFactory.ts]: Same as above

File: src/plugins/func/azureFunc.ts (hunk 8 above)
Comment (tbarlow12) [reply src/plugins/func/azureFunc.ts]: I'll put a try catch here

File: src/plugins/func/azureFunc.ts (hunk 7 above)
Comment (tbarlow12) [reply src/plugins/func/azureFunc.ts]: What other lifecycle are you referring to? This will just exit the process, and they would need to run again with a name specified

File: README.md [hunk 11]
```diff
@@ -18,6 +18,26 @@ This plugin enables Azure Functions support within the Serverless Framework.
...
 
+### Creating or removing Azure Functions
+
+To create a new Azure Function within your function app, run the following command from within your app's directory:
+
+```bash
+sls func add -n {functionName}
+```
+
+This will create a new `{functionName}` directory at the root of your application with `index.js` and `function.json` inside the directory. It will also update `serverless.yml` to contain the new function.
```
[the remark this answers is not in the document]
Comment (tbarlow12) [reply README.md]: As I've learned more about the tool, I don't know that we need a sample.dat. It's pretty basic and kind of confusing as to what it actually does. Was going to talk to the team about removing it from the template

File: src/plugins/func/azureFunc.ts (hunk 7 above)
Comment (mydiemho) [reply src/plugins/func/azureFunc.ts]: ah you're right, I was thinking of other command with more lifecyles

File: src/plugins/func/funcUtils.ts (hunk 9 above)
Comment (mydiemho) [reply src/plugins/func/funcUtils.ts]: yes but the plugin will generate incorrect `scriptFilePath` since `function.json` now resides at the same level as `index.json`

File: src/test/mockFactory.ts (hunk 10 above)
Comment (mydiemho) [reply src/test/mockFactory.ts]: same as above :p

File: README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (mydiemho) [reply README.md]: I have no strong preference, whatever you think is best :D

File: src/plugins/func/funcUtils.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    return JSON.stringify(httpBinding, null, 2);
+  }
+
+  public static getFunctionSlsObject(name: string, options: any) {
+    return FuncPluginUtils.defaultFunctionSlsObject(name);
+  }
+
+  private static defaultFunctionSlsObject(name: string) {
+    return {
+      handler: `index.handler`,
```
Comment (mydiemho) [inline src/plugins/func/funcUtils.ts]: Nit: can change back to double quote since we're no longer interpolating a variable

File: src/shared/utils.ts [hunk 13]
```diff
@@ -110,4 +111,11 @@ export class Utils {
   
     return metaData;
   }
+
+  public static interpolateFile(sls: Serverless, path: string, params: Map<string, string>) {
+    const template = sls.utils.readFileSync(path);
+    const names = params.keys();
+    const vals = params.values();
+    return new Function(...names, `return \`${template}\`;`)(...vals);
```
Comment (mydiemho) [inline src/shared/utils.ts]: Q: can 
```
`return \`${template}\`;`
```
be replace with 
```
`return ${template};`
```

Comment (tbarlow12) [reply src/shared/utils.ts]: This is actually a string interpolation that won't happen here. We actually have the same function (kind of) in VoTT. This is expecting a file with templated variables. I'm actually not using this right now... I was using it when I had a javascript template file. But I might keep it in just in case we go back to that

File: src/plugins/func/funcUtils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (PIC123) [reply src/plugins/func/funcUtils.ts]: Yah, the ideal solution would be to be able to include a js file of the handler since that's technically what we're packaging, but if that isn't really an option, then inline seems reasonable.

File: src/plugins/func/azureFunc.test.ts [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+      }, {createCwd: true, createTmp: true});    
+    });
+  
+    afterAll(() => {
+      mockFs.restore();
+    });
+  
+    it("returns with missing name", async () => {
+      const sls = MockFactory.createTestServerless();
+      const options = MockFactory.createTestServerlessOptions();
```
Comment (PIC123) [inline src/plugins/func/azureFunc.test.ts]: Q: Would it make sense to add a default name to the options and then remove/override as needed? Just thinking if we'll need it for other function tests if we add more functionality, like the option to include the extension bundles

File: src/plugins/func/funcUtils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (wbreza) [reply src/plugins/func/funcUtils.ts]: Let's add a work item to re-review this later.

File: src/plugins/func/azureFunc.test.ts (hunk 14 above)
Comment (tbarlow12) [reply src/plugins/func/azureFunc.test.ts]: For sure. Let's add that in another PR

File: src/plugins/func/funcUtils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tbarlow12) [reply src/plugins/func/funcUtils.ts]: [AB#18775]
